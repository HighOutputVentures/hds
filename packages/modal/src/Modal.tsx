import React, { FC, ReactNode } from 'react';
import { Button } from '@highoutput/hds';
import {
  As,
  Modal as ChakraModal,
  ModalOverlay,
  ModalContent,
  ModalHeader,
  ModalFooter,
  ModalBody,
  ModalCloseButton,
  Box,
  Flex,
  Icon,
  Spacer,
} from '@chakra-ui/react';

export interface ModalProps {
  isOpen: boolean;
  onClose: () => void;
  onOk: () => void;
  okText?: string;
  closeText?: string;
  size?: 'lg' | 'md' | 'sm' | 'xs';
  align?: 'left' | 'center' | 'right';
  icon?: As<any>;
  title?: ReactNode;
  children?: ReactNode;
}

const sizes = {
  xs: 'sm',
  sm: 'lg',
  md: '3xl',
  lg: '6xl',
};

const Modal: FC<ModalProps> = ({
  isOpen,
  onClose,
  onOk,
  okText = 'Ok',
  closeText = 'Close',
  size = 'sm',
  align,
  icon,
  title,
  children,
}) => {
  return (
    <ChakraModal isOpen={isOpen} onClose={onClose} size={sizes[size]}>
      <ModalOverlay />
      <ModalContent textAlign={align}>
        {icon && (
          <Box pl="4" pt="4" {...(align === 'right' && { mr: 5, mt: 7 })}>
            <Icon as={icon} width="48px" height="48px" />
          </Box>
        )}
        <Box flex={1}>
          <ModalCloseButton />
          <ModalHeader>{title}</ModalHeader>
          <ModalBody>{children}</ModalBody>
          <ModalFooter>
            <Flex w="full" justifyContent="flex-end">
              <Button onClick={onClose} block variant="outline-primary">
                {closeText}
              </Button>
              <Spacer maxW="4" />
              <Button onClick={onOk} block variant="solid-primary">
                {okText}
              </Button>
            </Flex>
          </ModalFooter>
        </Box>
      </ModalContent>
    </ChakraModal>
  );
};

export default Modal;
