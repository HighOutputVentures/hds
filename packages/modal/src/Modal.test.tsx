import { render } from '@testing-library/react';
import Modal from './Modal';
import * as React from 'react';

describe('Modal', () => {
  it('should render', () => {
    const { container } = render(
      <Modal isOpen onClose={jest.fn} onOk={jest.fn} title="Vans">
        Off the wall
      </Modal>
    );

    expect(container).toMatchSnapshot();
  });

  it('should render the correct title passed', () => {
    const title = 'This is a modal title';

    const { getByText } = render(
      <Modal isOpen onClose={jest.fn} onOk={jest.fn} title={title} />
    );

    expect(getByText(title)).toBeTruthy();
  });

  it('should render the correct body content', () => {
    const title = 'Hello! My name is Gustavo Fring';

    const { getByText } = render(
      <Modal isOpen onClose={jest.fn} onOk={jest.fn} title={title}>
        But you can call me gus.
      </Modal>
    );

    expect(getByText('But you can call me gus.')).toBeTruthy();
  });

  it('should render the correct text passed in close and ok button', () => {
    const title = 'Hello! My name is Gustavo Fring';
    const closeText = 'Dismiss';
    const okText = 'Accept';

    const { getByText } = render(
      <Modal
        isOpen
        onClose={jest.fn}
        onOk={jest.fn}
        title={title}
        closeText={closeText}
        okText={okText}
      >
        But you can call me gus.
      </Modal>
    );

    expect(getByText(closeText)).toBeTruthy();
    expect(getByText(okText)).toBeTruthy();
  });
});
