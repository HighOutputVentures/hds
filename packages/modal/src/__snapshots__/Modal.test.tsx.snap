// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal should render 1`] = `
<div
  aria-hidden="true"
  data-aria-hidden="true"
/>
`;
